assets/js/modules/analytics-4/utils/data-mock.js

```diff
@@ -0,0 +1,412 @@
+				case 'PERCENT':
+					values.push( {
+						value: faker.datatype
+							.float( { min: 0, max: 100 } )
+							.toString(),
+					} );
+					break;
+				case 'TIME':
+					values.push( {
+						value: faker.datatype
+							.number( { min: 0, max: 3600 } )
+							.toString(),
+					} ); // 1 hour max.
+					break;
+				case 'CURRENCY':
+					values.push( {
+						value: faker.datatype
+							.float( { min: 0, max: 10000 } )
+							.toString(),
+					} ); // $10k max.
+					break;
+			}
+		} );
+
+		metrics.push( ...values );
+	}
+
+	return metrics;
+}
+
+/**
+ * Sorts report rows and returns it.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Array.<Object>}        rows    Array of rows to sort.
+ * @param {Array.<Object>}        metrics Array of report metrics.
+ * @param {Object|Array.<Object>} orderby Sorting options.
+ * @return {Array.<Object>} Sorted rows.
+ */
+function sortRows( rows, metrics, orderby ) {
+	let sorted = rows;
+
+	const orders = castArray( orderby );
+	for ( const order of orders ) {
+		const direction = order?.sortOrder === 'DESCENDING' ? -1 : 1;
+		const index = metrics.findIndex(
+			( metric ) => getMetricKey( metric ) === order?.fieldName
+		);
+		if ( index < 0 ) {
+			continue;
+		}
+
+		sorted = sorted.sort( ( a, b ) => {
+			let valA = parseFloat( a.metrics[ index ]?.values?.[ 0 ] );
+			if ( Number.isNaN( valA ) ) {
+				valA = 0;
+			}
+
+			let valB = parseFloat( b.metrics[ index ]?.values?.[ 0 ] );
+			if ( Number.isNaN( valB ) ) {
+				valB = 0;
+			}
+
+			return ( valA - valB ) * direction;
+		} );
+	}
+
+	return sorted;
+}
+
+/**
+ * Generates mock data for Analytics 4 reports.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object} args Report options.
+ * @return {Array.<Object>} An array with generated report.
+ */
+export function getAnalytics4MockResponse( args ) {
+	invariant(
+		isPlainObject( args ),
+		'report options are required to generate a mock response.'
+	);
+	invariant(
+		isValidDateString( args.startDate ),
+		'a valid startDate is required.'
+	);
+	invariant(
+		isValidDateString( args.endDate ),
+		'a valid endDate is required.'
+	);
+
+	const originalSeedValue = faker.seedValue;
+	const argsHash = parseInt(
+		md5( args.url || 'http://example.com' ).substring( 0, 8 ),
+		16
+	);
+
+	// We set seed for every data mock to make sure that the same arguments get the same report data.
+	// It means that everyone will have the same report data and will see the same widgets in the storybook.
+	// This approach gives us additional flexibility to control randomness on a per widget basis.
+	if ( ! Number.isNaN( argsHash ) ) {
+		faker.seed( argsHash );
+	}
+
+	const data = {
+		rowCount: 0,
+		rows: [],
+		totals: [],
+		minimums: [],
+		maximums: [],
+		metadata: {
+			currencyCode: 'USD',
+			timeZone: 'America/Los_Angeles',
+		},
+		kind: 'analyticsData#runReport',
+	};
+
+	const { compareStartDate, compareEndDate } = args;
+	const metricValuesCount = compareStartDate && compareEndDate ? 2 : 1;
```

Comment:
The usage of `metricValuesCount` and related logic needs to be changed (and it should probably be renamed), due to the difference in response shape for GA4 whereby there are additional rows for each date range, rather than additional entries per-row. Also, and on a related note, the aggregate properties have a different structure. Please see the example GA4 response in this https://github.com/google/site-kit-wp/issues/6172#issuecomment-1375638193 for details.

Reply:
Thanks @techanvil . This is ready for you to review again.